compiler/rustc_hir_analysis/src/check/intrinsicck.rs

```diff
@@ -299,6 +299,28 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
             }
         }
 
+        match *ty.kind() {
+            ty::RawPtr(_, hir::Mutability::Mut) if asm.options.contains(InlineAsmOptions::READONLY) =>
+                self
+                    .tcx
+                    .dcx()
+                    .struct_span_warn(expr.span, "passing a mutable pointer to asm! block with 'readonly' option.")
+                    .with_note("`readonly` means that no memory write happens inside the asm! block.")
+                    .with_note("This is not limited to global variables, it also includes passed pointers.")
+                    .with_note("If passing this mutable pointer is intentional, remove the `readonly` attribute.")
+                    .emit(),
+            ty::RawPtr(_, _) if asm.options.contains(InlineAsmOptions::NOMEM) =>
+                self
+                    .tcx
+                    .dcx()
+                    .struct_span_warn(expr.span, "passing a pointer to asm! block with 'nomem' option.")
+                    .with_note("`nomem` means that no memory write or read happens inside the asm! block.")
+                    .with_note("This is not limited to global variables, it also includes passed pointers.")
+                    .with_note("If passing this pointer is intentional, replace the `nomem` attribute with `readonly` or remove it completely.")
```

Comment:
Same thing here, if passing the pointer is intentional then our recommendation should not involve changing the semantics of `asm!`.

Reply:
https://github.com/rust-lang/rust/pull/127063#discussion_r1674508742
> The problem here is that I don't see an obvious way to suppress the warning. Perhaps this lint would be better in clippy instead rather than something that generally applies to all Rust code?

At first, I wanted to make it an error, but probably some code would break and `asm!` is stable (could we jump on the edition2024 bandwagon?). Yeah, this is unsafe, and we can do a lot of stuff inside the `asm!` block, including freestyle transmute, however transmute does one job and *it* is dangerous, we can't do much with it besides "being smart" and looking for surrounding context, which is what clippy does with `eager_transmute` for example.

Here we could impose impose a simple rule that would help avoiding bugs with unintentional `nomem`. I can agree that the case with `*mut T` and `readonly` could be allowed, because we could pretend that the pointer is implicitly cast to `*const T` (like when calling functions that take `*const T`) and also the difference between `*mut` and `*const` is mostly cosmetic, but I don't see the same applying to `nomem` + `*const/mut T`.

Why would someone pass a pointer and do nothing with the data? If they need just the address, maybe they also need to consider if the provenance should be exposed or not. Idk, I think you know better than me how `asm!` is used, so if the final decision will be to not implement it here, then I won't insist. :thinking:

Reply:
just noting that there's like two, maybe three versions of "intended/intention" floating around in this discussion. let me see if I can name all of them:

- "I intended to pass `*mut T` + nomem: I will basically `ptr.addr()` it or something."
- "I intended to pass `*mut T` but I meant readonly: just `.cast_const()` it."
- "I intended to pass `*mut T`, and I'm totally gonna read it, lol, GIVE ME DAT UB!!! (no plz don't)"